### PR TITLE
multicompiler: add multicompiler package

### PR DESCRIPTION
**Pull request by kquick**
###### Motivation for this change

Adds the multicompiler, which is a fork of the llvm-based clang which provides artificial software diversity to protect software from code-reuse attacks.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by kquick** [comment]
Please note that the checks fail due to the Travis timeout on the build job; building llvm+clang takes longer than Travis will allow, but I have built this successfully in a local sandbox... several times, unfortunately :-).

**Comment by dtzWill** [comment]
Can you update this for latest master?

I'll review this shortly...

Also pinging cc-wrapper maintainers @Erickson2314 @orivej

File: pkgs/development/compilers/multicompiler/default.nix [hunk 1]
```diff
@@ -0,0 +1,140 @@
+{ pkgs ? import <nixpkgs> {}
```
Comment (dtzWill) [inline pkgs/development/compilers/multicompiler/default.nix]: Remove this please, add dependencies as arguments instead.

File: pkgs/development/compilers/multicompiler/default.nix [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      mv ${mc_llvm_data.name}-${mc_llvm_data.ver}-src multicompiler
+      chmod -R ug+rw multicompiler
+      cd multicompiler/tools
+      unpackFile ${fetch mc_clang_data}
+      mv ${mc_clang_data.name}-${mc_clang_data.ver}-src clang
+      chmod -R ug+rw clang
+      cd ..
+      '';
+
+    buildInputs = with pkgs; [ cmake libedit libxml2 llvm python perl groff libffi ];
```
Comment (dtzWill) [inline pkgs/development/compilers/multicompiler/default.nix]: cmake, python, perl, and groff should be "nativeBuildInputs".

File: pkgs/development/compilers/multicompiler/default.nix [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    (lib.optional pkgs.stdenv.isLinux "-DGCC_INSTALL_PREFIX=${gcc}") ++
+    (lib.optional (pkgs.stdenv.cc.libc != null) "-DC_INCLUDE_DIRS=${cc.libc}/include");
+
+    patches =
+      pkgs.stdenv.lib.optionals (!pkgs.stdenv.isDarwin) [./fix-llvm-config.patch ];
+
+    postPatch = ''
+      #sed -i -e 's/Args.hasArg(options::OPT_nostdlibinc)/true/' lib/Driver/Tools.cpp
+      #sed -i -e 's/DriverArgs.hasArg(options::OPT_nostdlibinc)/true/' lib/Driver/ToolChains.cpp
+    '';
```
Comment (dtzWill) [inline pkgs/development/compilers/multicompiler/default.nix]: Please look at current LLVM/Clang 3.5 expressions and sync this to match updates/fixes there.

File: pkgs/development/compilers/multicompiler/default.nix [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+      # install -m644 lib/clang/3.5.2/include/*.h $incdir/
+      mkdir -p $(dirname $incdir)
+      ln -s $out/lib/clang/3.5.2/include $incdir
+      set +x
+      '';
+
+    # Clang expects to find LLVMgold in its own prefix
+    # Clang expects to find sanitizer libraries in its own prefix
+    postInstall = ''
+      #ln -sv ${pkgs.llvm}/lib/LLVMgold.so $out/lib
```
Comment (dtzWill) [inline pkgs/development/compilers/multicompiler/default.nix]: What's this about?

**Review by dtzWill — CHANGES_REQUESTED** [review]

**Comment by Ericson2314** [comment]
Oh! I am sorry I didn't see this earlier. I have in fact used the multicompiler.

**Comment by Ericson2314** [comment]
This had looked pretty good modulo @dtzWill's comments. Now that we support cross compilation with llvm, it should be even easier.

**Pull request closed without merging**